packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,190 @@
+      for (const runId of runs) {
+        const state = stateByRunId.get(runId);
+        if (state) {
+          states.push(state);
+        } else {
+          const runState = await this.getSingleState(runId, extractionClientApi, accessToken);
+          states.push(runState);
+          stateByRunId.set(runId, runState);
+        }
+      }
+      const finalState = this.getFinalState(states);
+      stateByReportId.set(reportId, finalState);
+    }
+    return stateByReportId;
+  }
+
+  public clearJobs() {
+    this._reportRunIds = new Map<string, string[]>();
+  }
+
+  private getFinalState(states: ExtractorState[]) {
+    const responseState = states.includes(ExtractorState.Failed) ? ExtractionStates.Failed :
+      states.includes(ExtractorState.Queued) ? ExtractionStates.Queued :
+        states.includes(ExtractorState.Running) ? ExtractionStates.Running :
+          states.includes(ExtractorState.Succeeded) ? ExtractionStates.Succeeded : ExtractionStates.Failed;
+    return responseState;
+  }
+
+  private async getSingleState(runId: string, extractionClientApi: ExtractionClient, accessToken: string): Promise<ExtractorState> {
+    try {
+      const response = await extractionClientApi.getExtractionStatus(accessToken, runId);
+      return response.state;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return ExtractorState.Failed;
+  }
+
+  public async startJobs(reportIds: string[]) {
+    const extractionClientApi = new ExtractionClient(
+      generateUrl(REPORTING_BASE_PATH, this._apiConfig.baseUrl)
+    );
+    const accessToken = await this._apiConfig.getAccessToken();
+
+    const extractionByIModel = new Map<string, string>();
+    for (const reportId of reportIds) {
+      const IModels = (await this.fetchReportMappings(reportId, this._apiConfig)).map((m) => m.imodelId);
+      const uniqueIModels = Array.from(new Set(IModels));
+
+      const runs: string[] = [];
+      for (const iModelId of uniqueIModels) {
+        const extraction = extractionByIModel.get(iModelId);
+        if (extraction) {
+          runs.push(extraction);
+        } else {
+          const run = await this.runExtraction(iModelId, extractionClientApi, accessToken);
+          if (run) {
+            runs.push(run.id);
+            extractionByIModel.set(iModelId, run.id);
+          }
+        }
+      }
+      this._reportRunIds.set(reportId, runs);
+    }
+  }
+
+  private async runExtraction(iModelId: string, extractionClientApi: ExtractionClient, accessToken: string) {
+    try {
+      const response = await extractionClientApi.runExtraction(
+        accessToken,
+        iModelId
+      );
+      return response;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return undefined;
+  }
+
+  private async fetchReportMappings(
+    reportId: string,
+    apiContext: ReportsApiConfig
+  ): Promise<ReportMappingAndMapping[]> {
+    try {
+      const reportsClientApi = new ReportsClient(
```

Comment:
Clients should be stored in a private field and reused.

Whenever baseUrl changes the BulkExtractor object should be recreated, because `_reportRunIds` becomes invalid.

Reply:
All clients were moved to a private field and are initialized in constructor